source/extensions/filters/http/common/compressor/compressor.cc

```diff
@@ -0,0 +1,412 @@
+namespace HttpFilters {
+namespace Common {
+namespace Compressors {
+
+namespace {
+
+// Default minimum length of an upstream response that allows compression.
+const uint64_t DefaultMinimumContentLength = 30;
+
+// Default content types will be used if any is provided by the user.
+const std::vector<std::string>& defaultContentEncoding() {
+  CONSTRUCT_ON_FIRST_USE(
+      std::vector<std::string>,
+      {"text/html", "text/plain", "text/css", "application/javascript", "application/x-javascript",
+       "text/javascript", "text/x-javascript", "text/ecmascript", "text/js", "text/jscript",
+       "text/x-js", "application/ecmascript", "application/x-json", "application/xml",
+       "application/json", "image/svg+xml", "text/xml", "application/xhtml+xml"});
+}
+
+// List of CompressorFilterConfig objects registered for a stream.
+struct CompressorRegistry : public StreamInfo::FilterState::Object {
+  std::list<CompressorFilterConfigSharedPtr> filter_configs_;
+};
+
+// Key to per stream CompressorRegistry objects.
+const std::string& compressorRegistryKey() { CONSTRUCT_ON_FIRST_USE(std::string, "compressors"); }
+
+} // namespace
+
+CompressorFilterConfig::CompressorFilterConfig(
+    const envoy::extensions::filters::http::compressor::v3::Compressor& compressor,
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+    const std::string& content_encoding)
+    : content_length_(contentLengthUint(compressor.content_length().value())),
+      content_type_values_(contentTypeSet(compressor.content_type())),
+      disable_on_etag_header_(compressor.disable_on_etag_header()),
+      remove_accept_encoding_header_(compressor.remove_accept_encoding_header()),
+      stats_(generateStats(stats_prefix, scope)), runtime_(runtime),
+      content_encoding_(content_encoding) {}
+
+StringUtil::CaseUnorderedSet
+CompressorFilterConfig::contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types) {
+  const auto& default_content_encodings = defaultContentEncoding();
+  return types.empty() ? StringUtil::CaseUnorderedSet(default_content_encodings.begin(),
+                                                      default_content_encodings.end())
+                       : StringUtil::CaseUnorderedSet(types.cbegin(), types.cend());
+}
+
+uint32_t CompressorFilterConfig::contentLengthUint(Protobuf::uint32 length) {
+  return length > 0 ? length : DefaultMinimumContentLength;
+}
+
+CompressorFilter::CompressorFilter(const CompressorFilterConfigSharedPtr config)
+    : skip_compression_{true}, config_(std::move(config)) {}
+
+Http::FilterHeadersStatus CompressorFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  if (accept_encoding != nullptr) {
+    accept_encoding_ = std::make_unique<std::string>(accept_encoding->value().getStringView());
+  }
+
+  if (config_->runtime().snapshot().featureEnabled(config_->featureName(), 100)) {
```

Comment:
Where does feature name come from here exactly? Is this just for purposes of having a kill switch? In new code we prefer an explicit kill switch on the config if needed. See `RuntimeFeatureFlag`

Reply:
This is a slightly modified copy-paste from the [gzip](https://github.com/envoyproxy/envoy/blob/bb7ceff4c3c5bd4555dff28b6e56d27f2f8be0a7/source/extensions/filters/http/gzip/gzip_filter.cc#L105) filter. I presume this is used for canary tests.

Probably `RuntimeFractionalPercent` would be a better candidate.

`featureName` is supposed to be a constant set in a class derived from `CompressorFilterConfig`, e.g. `GzipFilterConfig`.

Reply:
Yeah I think this is all before we had the explicit `RuntimeFeatureFlag`. Can you add this to the new config and use that instead if we are doing a new thing? That will be better for the future and we can remove this explicit `featureName()` variable as a hard coded thing. Thank you!